src/core/src/package_managers/UbuntuProClient.py

```diff
@@ -42,26 +47,118 @@ def is_pro_working(self):
+        self.composite_logger.log_debug("Is Ubuntu Pro Client working debug flags: [Success={0}][UbuntuProClientVersion={1}][UbuntuProClientMinimumVersionInstalled={2}][IsAttached={3}][Error={4}]".format(is_ubuntu_pro_client_working, ubuntu_pro_client_version, is_minimum_ubuntu_pro_version_installed, self.ubuntu_pro_client_is_attached, ubuntu_pro_client_exception))
         return is_ubuntu_pro_client_working
 
+    def log_ubuntu_pro_client_attached(self):
+        """log the attachment status of the machine."""
+        ubuntu_pro_client_is_attached = False
+        try:
+            code, output = self.env_layer.run_command_output(self.ubuntu_pro_client_security_status_cmd, False, False)
+            if code == 0:
+                ubuntu_pro_client_is_attached = json.loads(output)['summary']['ua']['attached']
+        except Exception as error:
+            ubuntu_pro_client_exception = repr(error)
+            self.composite_logger.log_debug("Ubuntu Pro Client Attached Exception: [Exception={0}]".format(ubuntu_pro_client_exception))
+        return ubuntu_pro_client_is_attached
+
+    def extract_packages_and_versions(self, updates):
+        extracted_updates = []
+        extracted_updates_versions = []
+
+        for update in updates:
+            extracted_updates.append(update.package)
+            if not self.ubuntu_pro_client_is_attached and update.provided_by in self.security_esm_criteria_strings:
+                extracted_updates_versions.append(Constants.UA_ESM_REQUIRED)
+            else:
+                extracted_updates_versions.append(update.version)
+        return extracted_updates, extracted_updates_versions
+
     def get_security_updates(self):
-        pass
+        """query Ubuntu Pro Client to get security updates."""
+        security_updates_query_success = False
+        security_updates = []
+        security_updates_versions = []
+        security_updates_exception = None
+        try:
+            ubuntu_pro_client_updates = self.get_ubuntu_pro_client_updates()
+            security_updates_query_success = True
+            security_criteria_string = ["standard-security"]
+            filtered_security_updates = [update for update in ubuntu_pro_client_updates if update.provided_by in security_criteria_string]
+            security_updates, security_updates_versions = self.extract_packages_and_versions(filtered_security_updates)
+        except Exception as error:
+            security_updates_exception = repr(error)
+
+        self.composite_logger.log_debug("Ubuntu Pro Client get security updates : [SecurityUpdatesCount={0}][error={1}]".format(len(security_updates), security_updates_exception))
+        return security_updates_query_success, security_updates, security_updates_versions
+
+    def get_security_esm_updates(self):
```

Comment:
There's a lot of common code between get_security_esm_updates() and get_security_updates(), we should refactor

Reply:
Also get_other_updates()

Reply:
Done